File: web/lib/irpf.ts
/** Tipos + helpers do módulo IRPF.
 *
 * Convenção:
 * - Leituras: Server Components consultam Supabase diretamente (RLS protege)
 * - Mutações: server actions chamam a API Go via fetchIrpf (com token do user logado)
 */

import { API_BASE_URL } from './api'

// ─── Tipos ──────────────────────────────────────────────────────────────────

export type IrpfStatus =
  | 'a_iniciar' | 'coletando' | 'em_processamento' | 'aguardando_cliente'
  | 'entregue' | 'em_malha' | 'retificada' | 'cancelada'

export type IrpfSituacaoFinal = 'a_restituir' | 'a_pagar' | 'sem_imposto'

export type IrpfLancamentoTipo =
  | 'rendimento_tributavel' | 'rendimento_isento' | 'rendimento_exclusivo'
  | 'deducao_medica' | 'deducao_educacao' | 'deducao_previdencia' | 'deducao_pensao'
  | 'dependente' | 'bem_direito' | 'divida'

export interface IrpfDeclarante {
  id: string
  org_id: string
  empresa_id?: string
  cpf: string
  nome_completo: string
  data_nascimento?: string
  email?: string
  telefone?: string
  observacoes?: string
  created_at: string
  updated_at: string
  declaracoes_count?: number
}

export interface IrpfDeclaracao {
  id: string
  org_id: string
  declarante_id: string
  declarante_nome?: string
  declarante_cpf?: string
  exercicio: number
  ano_calendario: number
  status: IrpfStatus
  responsavel_id?: string
  responsavel_nome?: string
  rendimentos_total_cents: number
  deducoes_total_cents: number
  imposto_devido_cents: number
  imposto_retido_cents: number
  saldo_cents: number
  situacao_final?: IrpfSituacaoFinal
  recibo_url?: string
  transmitida_em?: string
  observacoes?: string
  created_at: string
  updated_at: string
}

export interface IrpfLancamento {
  id: string
  org_id: string
  declaracao_id: string
  tipo: IrpfLancamentoTipo
  fonte_pagadora?: string
  fonte_cnpj?: string
  descricao?: string
  valor_cents: number
  imposto_retido_cents: number
  documento_url?: string
  payload: Record<string, unknown>
  created_at: string
}

export interface IrpfDeclaracaoDetalhe {
  declaracao: IrpfDeclaracao
  declarante: IrpfDeclarante
  lancamentos: IrpfLancamento[]
}

export interface IrpfDashboard {
  exercicio: number
  total: number
  a_iniciar: number
  coletando: number
  em_processamento: number
  aguardando_cliente: number
  entregues: number
  em_malha: number
  retificadas: number
  canceladas: number
  total_a_restituir_cents: number
  total_a_pagar_cents: number
  total_imposto_retido_cents: number
}

// ─── Helpers ────────────────────────────────────────────────────────────────

export async function fetchIrpf<T>(token: string, path: string, init: RequestInit = {}): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
      ...(init.headers ?? {}),
    },
    cache: 'no-store',
  })
  const text = await res.text()
  const body = text ? safeJson(text) : null
  if (!res.ok) {
    const msg = (body as { message?: string } | null)?.message ?? `${res.status} ${res.statusText}`
    throw new Error(msg)
  }
  return body as T
}

function safeJson(t: string): unknown {
  try { return JSON.parse(t) } catch { return t }
}

// ─── Formatadores e labels ──────────────────────────────────────────────────

export const statusLabel: Record<IrpfStatus, string> = {
  a_iniciar: 'A iniciar',
  coletando: 'Coletando',
  em_processamento: 'Em processamento',
  aguardando_cliente: 'Aguardando cliente',
  entregue: 'Entregue',
  em_malha: 'Em malha',
  retificada: 'Retificada',
  cancelada: 'Cancelada',
}

export const statusBadge: Record<IrpfStatus, string> = {
  a_iniciar: 'bg-ink-100 text-ink-700 ring-ink-200',
  coletando: 'bg-amber-100 text-amber-900 ring-amber-300',
  em_processamento: 'bg-blue-100 text-blue-900 ring-blue-300',
  aguardando_cliente: 'bg-purple-100 text-purple-900 ring-purple-300',
  entregue: 'bg-emerald-100 text-emerald-900 ring-emerald-300',
  em_malha: 'bg-rose-100 text-rose-900 ring-rose-300',
  retificada: 'bg-ink-100 text-ink-700 ring-ink-200',
  cancelada: 'bg-ink-100 text-ink-500 ring-ink-200',
}

export const tipoLabel: Record<IrpfLancamentoTipo, string> = {
  rendimento_tributavel: 'Rendimento tributável',
  rendimento_isento: 'Rendimento isento',
  rendimento_exclusivo: 'Rendimento exclusivo na fonte',
  deducao_medica: 'Dedução médica',
  deducao_educacao: 'Dedução educação',
  deducao_previdencia: 'Dedução previdência',
  deducao_pensao: 'Dedução pensão alimentícia',
  dependente: 'Dependente',
  bem_direito: 'Bem ou direito',
  divida: 'Dívida / ônus real',
}

export function exercicioAtual(): number {
  // Exercício do IRPF = ano de entrega = ano-calendário + 1
  const d = new Date()
  return d.getMonth() < 4 ? d.getFullYear() : d.getFullYear() + 1
}

export function formatCPF(cpf: string): string {
  const digits = (cpf ?? '').replace(/\D/g, '')
  if (digits.length !== 11) return cpf
  return `${digits.slice(0, 3)}.${digits.slice(3, 6)}.${digits.slice(6, 9)}-${digits.slice(9, 11)}`
}

File: api/internal/models/irpf.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ─── Declarante ────────────────────────────────────────────────────────────
type IrpfDeclarante struct {
	ID             uuid.UUID  `json:"id"`
	OrgID          uuid.UUID  `json:"org_id"`
	EmpresaID      *uuid.UUID `json:"empresa_id,omitempty"`
	CPF            string     `json:"cpf"`
	NomeCompleto   string     `json:"nome_completo"`
	DataNascimento *time.Time `json:"data_nascimento,omitempty"`
	Email          *string    `json:"email,omitempty"`
	Telefone       *string    `json:"telefone,omitempty"`
	Observacoes    *string    `json:"observacoes,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`

	// Agregado opcional preenchido por listagem
	DeclaracoesCount int `json:"declaracoes_count,omitempty"`
}

type CreateIrpfDeclaranteDTO struct {
	EmpresaID      *uuid.UUID `json:"empresa_id"`
	CPF            string     `json:"cpf"`
	NomeCompleto   string     `json:"nome_completo"`
	DataNascimento *string    `json:"data_nascimento"`
	Email          *string    `json:"email"`
	Telefone       *string    `json:"telefone"`
	Observacoes    *string    `json:"observacoes"`
}

type UpdateIrpfDeclaranteDTO struct {
	EmpresaID      *uuid.UUID `json:"empresa_id"`
	NomeCompleto   *string    `json:"nome_completo"`
	DataNascimento *string    `json:"data_nascimento"`
	Email          *string    `json:"email"`
	Telefone       *string    `json:"telefone"`
	Observacoes    *string    `json:"observacoes"`
}

// ─── Declaração ────────────────────────────────────────────────────────────
type IrpfDeclaracao struct {
	ID                    uuid.UUID  `json:"id"`
	OrgID                 uuid.UUID  `json:"org_id"`
	DeclaranteID          uuid.UUID  `json:"declarante_id"`
	DeclaranteNome        *string    `json:"declarante_nome,omitempty"`
	DeclaranteCPF         *string    `json:"declarante_cpf,omitempty"`
	Exercicio             int        `json:"exercicio"`
	AnoCalendario         int        `json:"ano_calendario"`
	Status                string     `json:"status"`
	ResponsavelID         *uuid.UUID `json:"responsavel_id,omitempty"`
	ResponsavelNome       *string    `json:"responsavel_nome,omitempty"`
	RendimentosTotalCents int64      `json:"rendimentos_total_cents"`
	DeducoesTotalCents    int64      `json:"deducoes_total_cents"`
	ImpostoDevidoCents    int64      `json:"imposto_devido_cents"`
	ImpostoRetidoCents    int64      `json:"imposto_retido_cents"`
	SaldoCents            int64      `json:"saldo_cents"`
	SituacaoFinal         *string    `json:"situacao_final,omitempty"`
	ReciboURL             *string    `json:"recibo_url,omitempty"`
	TransmitidaEm         *time.Time `json:"transmitida_em,omitempty"`
	Observacoes           *string    `json:"observacoes,omitempty"`
	CreatedAt             time.Time  `json:"created_at"`
	UpdatedAt             time.Time  `json:"updated_at"`
}

type CreateIrpfDeclaracaoDTO struct {
	DeclaranteID  uuid.UUID `json:"declarante_id"`
	Exercicio     int       `json:"exercicio"`
	AnoCalendario int       `json:"ano_calendario"`
	ResponsavelID *uuid.UUID `json:"responsavel_id"`
	Observacoes   *string    `json:"observacoes"`
}

type UpdateIrpfDeclaracaoDTO struct {
	Status        *string    `json:"status"`
	ResponsavelID *uuid.UUID `json:"responsavel_id"`
	Observacoes   *string    `json:"observacoes"`
	ReciboURL     *string    `json:"recibo_url"`
}

// ─── Lançamento ────────────────────────────────────────────────────────────
type IrpfLancamento struct {
	ID                 uuid.UUID `json:"id"`
	OrgID              uuid.UUID `json:"org_id"`
	DeclaracaoID       uuid.UUID `json:"declaracao_id"`
	Tipo               string    `json:"tipo"`
	FontePagadora      *string   `json:"fonte_pagadora,omitempty"`
	FonteCNPJ          *string   `json:"fonte_cnpj,omitempty"`
	Descricao          *string   `json:"descricao,omitempty"`
	ValorCents         int64     `json:"valor_cents"`
	ImpostoRetidoCents int64     `json:"imposto_retido_cents"`
	DocumentoURL       *string   `json:"documento_url,omitempty"`
	Payload            map[string]any `json:"payload"`
	CreatedAt          time.Time `json:"created_at"`
}

type CreateIrpfLancamentoDTO struct {
	Tipo               string         `json:"tipo"`
	FontePagadora      *string        `json:"fonte_pagadora"`
	FonteCNPJ          *string        `json:"fonte_cnpj"`
	Descricao          *string        `json:"descricao"`
	ValorCents         int64          `json:"valor_cents"`
	ImpostoRetidoCents int64          `json:"imposto_retido_cents"`
	DocumentoURL       *string        `json:"documento_url"`
	Payload            map[string]any `json:"payload"`
}

// ─── Detalhe consolidado de uma declaração (resposta de GET /irpf/declaracoes/:id) ─
type IrpfDeclaracaoDetalhe struct {
	Declaracao  IrpfDeclaracao    `json:"declaracao"`
	Declarante  IrpfDeclarante    `json:"declarante"`
	Lancamentos []IrpfLancamento  `json:"lancamentos"`
}

// ─── Dashboard do exercício ────────────────────────────────────────────────
type IrpfDashboard struct {
	Exercicio          int   `json:"exercicio"`
	Total              int   `json:"total"`
	AIniciar           int   `json:"a_iniciar"`
	Coletando          int   `json:"coletando"`
	EmProcessamento    int   `json:"em_processamento"`
	AguardandoCliente  int   `json:"aguardando_cliente"`
	Entregues          int   `json:"entregues"`
	EmMalha            int   `json:"em_malha"`
	Retificadas        int   `json:"retificadas"`
	Canceladas         int   `json:"canceladas"`
	TotalARestituirCents int64 `json:"total_a_restituir_cents"`
	TotalAPagarCents     int64 `json:"total_a_pagar_cents"`
	TotalImpostoRetidoCents int64 `json:"total_imposto_retido_cents"`
}
